imputation_sklearn optimization

dattranm opened the pull request.

VishnuGowthem commented on src/main/anovos/data_transformer/transformers.py:
```diff
@@ -1763,6 +1802,8 @@ def imputation_sklearn(
     """
 
+    if persist:
+        idf = idf.persist(persist_option)
```
this is more generically for persist_option - not at this point in code I mean.

dattranm replied:
In here, whenever `pyspark.StorageLevel` receive the wrong value, the pyspark library has its own builtin error check for this:

dattranm commented on src/main/anovos/data_transformer/transformers.py:
```diff
@@ -1676,17 +1677,24 @@ def imputation_MMM(
-    pre_existing_model=False,
-    model_path="NA",
-    output_mode="replace",
+    list_of_cols: list = "missing",
```
I removed the suggested `list` type for `list_of_cols` Vishnu, and add a bit more description to explain this.
For the second point, across the repo, in `imputation_MMM` function, we also use the same `list_of_cols="missing"`. I think we should stick with this for now to standardize the argument.

dattranm commented on src/main/anovos/data_transformer/transformers.py:
```diff
@@ -1731,9 +1740,39 @@ def imputation_sklearn(
+        It is recommended to set this as True for large datasets. (Default value = True)
+    persist_option
+        If persist is True, this argument is used to determine the type of persist.
+        (Default value = pyspark.StorageLevel.MEMORY_AND_DISK)
```
I have added the list of the pyspark.StorageLevel from pyspark official documentation in here.
For the label, according to Killian comment on #268 ,https://github.com/anovos/anovos/pull/268#discussion_r948473258, it would be better to just set it to pyspark.StorageLevel.MEMORY_AND_DISK to avoid confusion.

VishnuGowthem replied:
okay noted.

dattranm commented on src/main/anovos/data_transformer/transformers.py:
```diff
@@ -1851,10 +1892,22 @@ def imputation_sklearn(
+                    idf_model = idf_model.persist(persist_option)
+        else:
+            idf_model = idf
         idf_pd = idf_model.select(list_of_cols).toPandas()
```
Resolved with https://github.com/anovos/anovos/pull/337/commits/402964046aaa383a06f71224499800d1a37f7f89

dattranm commented on src/main/anovos/data_transformer/transformers.py:
```diff
@@ -1676,17 +1677,24 @@ def imputation_MMM(
+    strata_cols: str = "all",
+    stratified_type: str = "population",
+    sample_size: int = 10000,
+    sample_seed: int = 42,
```
got it Vishnu. I will keep it like this for now, as you can see the same usage of `data_sample` in `correlation_matrix` function.
https://github.com/anovos/anovos/blob/a833099c9ec6dc2ef3ee68e2fca638bc4eaa8379/src/main/anovos/data_analyzer/association_evaluator.py#L40

dattranm commented on src/main/anovos/data_transformer/transformers.py:
```diff
@@ -1731,9 +1740,39 @@ def imputation_sklearn(
+        (Default value = "regression")
+    use_sampling
+        Boolean argument - True or False. This argument is used to determine whether to use random sample method on
+        source and target dataset, True will enable the use of sample method, otherwise False.
```
Resolved with https://github.com/anovos/anovos/commit/402964046aaa383a06f71224499800d1a37f7f89

dattranm commented:
Thank @ranjanravish and @VishnuGowthem for the comment and review! I have addressed all of your comments, with some comments of my own that can lead to further discussion. Can you guys have a quick look again and let me know if there is anything else I need to address.

VishnuGowthem reviewed (approved):
I am good with the changes subject to all workflows passing, thanks @dattranm !

VishnuGowthem commented:
> I am good with the changes subject to all workflows passing, thanks @dattranm !

I think last set of workflows has passed. So I am good to merge this. Thanks !

dattranm commented:
Thanks @VishnuGowthem for your review and approval. @kajanansangar can you have a final look and review on this to confirm everything is ready to merge?
